File: ios/liquid_ai/Sources/liquid_ai/GenerationProgressHandler.swift
import Flutter
import Foundation

/// Handles streaming generation events to Flutter via EventChannel.
final class GenerationProgressHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private let lock = NSLock()

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        lock.lock()
        defer { lock.unlock() }
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        lock.lock()
        defer { lock.unlock() }
        eventSink = nil
        return nil
    }

    /// Sends a text chunk event.
    func sendChunk(generationId: String, chunk: String) {
        sendEvent([
            "generationId": generationId,
            "type": "chunk",
            "chunk": chunk
        ])
    }

    /// Sends a reasoning chunk event.
    func sendReasoningChunk(generationId: String, chunk: String) {
        sendEvent([
            "generationId": generationId,
            "type": "reasoningChunk",
            "chunk": chunk
        ])
    }

    /// Sends an audio sample event.
    func sendAudioSamples(
        generationId: String,
        samples: [Float],
        sampleRate: Int
    ) {
        sendEvent([
            "generationId": generationId,
            "type": "audioSample",
            "audioSamples": samples,
            "sampleRate": sampleRate
        ])
    }

    /// Sends a function call event.
    func sendFunctionCalls(
        generationId: String,
        calls: [[String: Any]]
    ) {
        sendEvent([
            "generationId": generationId,
            "type": "functionCall",
            "functionCalls": calls
        ])
    }

    /// Sends a completion event.
    func sendComplete(
        generationId: String,
        message: [String: Any],
        finishReason: String,
        stats: [String: Any]?
    ) {
        var event: [String: Any] = [
            "generationId": generationId,
            "type": "complete",
            "message": message,
            "finishReason": finishReason
        ]
        if let stats = stats {
            event["stats"] = stats
        }
        sendEvent(event)
    }

    /// Sends an error event.
    func sendError(generationId: String, error: String) {
        sendEvent([
            "generationId": generationId,
            "type": "error",
            "error": error
        ])
    }

    /// Sends a cancellation event.
    func sendCancelled(generationId: String) {
        sendEvent([
            "generationId": generationId,
            "type": "cancelled"
        ])
    }

    private func sendEvent(_ event: [String: Any]) {
        lock.lock()
        let sink = eventSink
        lock.unlock()

        DispatchQueue.main.async {
            sink?(event)
        }
    }
}

File: android/src/main/kotlin/dev/sogl/liquid_ai/GenerationProgressHandler.kt
package dev.sogl.liquid_ai

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/// Handles streaming generation events to Flutter via EventChannel.
class GenerationProgressHandler : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val lock = Any()

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        synchronized(lock) {
            eventSink = events
        }
    }

    override fun onCancel(arguments: Any?) {
        synchronized(lock) {
            eventSink = null
        }
    }

    /// Sends a text chunk event.
    fun sendChunk(generationId: String, chunk: String) {
        sendEvent(mapOf(
            "generationId" to generationId,
            "type" to "chunk",
            "chunk" to chunk
        ))
    }

    /// Sends a reasoning chunk event.
    fun sendReasoningChunk(generationId: String, chunk: String) {
        sendEvent(mapOf(
            "generationId" to generationId,
            "type" to "reasoningChunk",
            "chunk" to chunk
        ))
    }

    /// Sends an audio sample event.
    fun sendAudioSamples(
        generationId: String,
        samples: List<Float>,
        sampleRate: Int
    ) {
        sendEvent(mapOf(
            "generationId" to generationId,
            "type" to "audioSample",
            "audioSamples" to samples,
            "sampleRate" to sampleRate
        ))
    }

    /// Sends a function call event.
    fun sendFunctionCalls(
        generationId: String,
        calls: List<Map<String, Any>>
    ) {
        sendEvent(mapOf(
            "generationId" to generationId,
            "type" to "functionCall",
            "functionCalls" to calls
        ))
    }

    /// Sends a completion event.
    fun sendComplete(
        generationId: String,
        message: Map<String, Any>,
        finishReason: String,
        stats: Map<String, Any>?
    ) {
        val event = mutableMapOf<String, Any>(
            "generationId" to generationId,
            "type" to "complete",
            "message" to message,
            "finishReason" to finishReason
        )
        stats?.let { event["stats"] = it }
        sendEvent(event)
    }

    /// Sends an error event.
    fun sendError(generationId: String, error: String) {
        sendEvent(mapOf(
            "generationId" to generationId,
            "type" to "error",
            "error" to error
        ))
    }

    /// Sends a cancellation event.
    fun sendCancelled(generationId: String) {
        sendEvent(mapOf(
            "generationId" to generationId,
            "type" to "cancelled"
        ))
    }

    private fun sendEvent(event: Map<String, Any>) {
        val sink: EventChannel.EventSink?
        synchronized(lock) {
            sink = eventSink
        }

        mainHandler.post {
            sink?.success(event)
        }
    }
}
